### PR TITLE
Infer column type two

### LINKED ISSUE #92: Type Inference 2: Determine a type for a column
**Problem**

Different types in Mathesar will enable different operations; for example, strings could be aggregated by concatenating, but numeric types could be aggregated by summing or multiplying. So far, while we can reflect different types, we have no way to determine the type most appropriate for a column.

**Proposed solution**

Using the function implemented for #91 , we should then test the column against types in an algorithmic way to determine the best type prediction we can for the column

### PR DESCRIPTION
**Pull request by mathemancer**
Fixes #92 

This adds a function to infer the type of the column by attempting to alter the column to various types in a logical order.

**Technical details**

This has the side effect of losing data in the case where the column is altered to a type via some non-injective mapping.  For example, altering the strings `'0123'` and `'123'` to a numeric will result in the value `123` in both cases.  It's not possible to recover the original strings from the resulting value.

We'll add mechanisms to recover the original version of an altered column in future PRs.

## Checklist
 
- [X] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [X] My pull request targets the `master` branch of the repository
- [X] My commit messages follow [best practices][best_practices].
- [X] My code follows the established code style of the repository.
- [X] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [X] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: db/tests/types/test_inference.py
Comment (kgodey) [inline db/tests/types/test_inference.py:18]: You're using this fixture in multiple files, I recommend putting it in `conftest.py` in the root of this folder so that you can reuse it.

Comment (kgodey) [reply db/tests/types/test_inference.py:18]: Same with other reused fixtures.

File: db/tests/types/test_inference.py
Comment (kgodey) [inline db/tests/types/test_inference.py:52]: Is this a debug print?

File: db/types/inference.py
Comment (kgodey) [inline db/types/inference.py:74]: Shouldn't this be `logger.error`?

**Review by kgodey — APPROVED** [review]
Approving to unblock, please make requested changes before merging.

File: db/tests/types/test_inference.py
Comment (mathemancer) [reply db/tests/types/test_inference.py:52]: oops, good eye.

File: db/tests/types/test_inference.py
Comment (mathemancer) [reply db/tests/types/test_inference.py:18]: My reasoning was that these fixtures use code from mathesar that is tested by test files with _no_ access to these fixtures.  (This makes sure we're not using fixtures in a way that invalidates the code they use.  Do you know of a way to separate those out (or make it difficult/obvious when you _shouldn't_ use a given fixture for testing?  In particular, I'd want to find some way to guarantee that these fixtures won't be used for testing the `db/types/install.py` file at some point in the future.

Comment (kgodey) [reply db/tests/types/test_inference.py:18]: You could organize the files differently so that any tests that can use these fixtures are in a single directory with the appropriate `conftest.py` and files that shouldn't use these fixtures are in a different directory.

Comment (kgodey) [reply db/tests/types/test_inference.py:18]: that seems like the obvious solution to me but I might be missing something.

File: db/types/inference.py
Comment (mathemancer) [reply db/types/inference.py:74]: This exception is an expected part of the control flow (it's how we know when a given type is inappropriate for a given column).  I think `info` is reasonable for that.  I _do_ think I should remove the spot where I set the log level to `INFO` in this file, though (it should be set by the importer if they want a non-default level).  What do you think?

File: db/tests/types/test_inference.py
Comment (mathemancer) [reply db/tests/types/test_inference.py:18]: We'd either need to break the "test file directory structure mirroring the real file directory structure" convention, or reorganize the real files in a (to me) less logical way, then mirror it.  I'm okay with the former, I guess, but I tend towards letting code repeat in test files over obfuscating what the files are testing (though it's not too hard to figure out at this stage in the project).

Comment (mathemancer) [reply db/tests/types/test_inference.py:18]: We could also put these fixtures in a non-default file (i.e., not a `conftest.py` file), then import it.  It's kind of non-standard, but I think it'd be less confusing than partially mirroring the directory structure of the tested files with the structure of the test files.

Comment (mathemancer) [reply db/tests/types/test_inference.py:18]: Actually, I think that's probably my favorite solution.

Comment (mathemancer) [reply db/tests/types/test_inference.py:18]: For that kind of reorganization, I think it'd merit its own PR to document what was done and why with the restructuring.  @kgodey, what do you think?

Comment (kgodey) [reply db/tests/types/test_inference.py:18]: @mathemancer I think importing a non-standard file makes sense, and doing it in a separate PR also makes sense. I was originally thinking you'd reorganize the real files as well, but I like your solution too.

Comment (mathemancer) [reply db/tests/types/test_inference.py:18]: I'll double-check that the organization of the real files makes sense and adjust accordingly.

File: db/types/inference.py
Comment (kgodey) [reply db/types/inference.py:68]: Can you catch the specific exception you're expecting and add a comment explaining it's the expected flow?

I agree re: removing log level in the file.

Comment (mathemancer) [reply db/types/inference.py:68]: Unfortunately, there's a bunch of different exceptions based on which type was casting to which other.  I can try to figure out if they're sub-classes of something appropriate (I think python will catch the parent class in that case, I'll have to check)

Comment (mathemancer) [reply db/types/inference.py:71]: Okay, we'll use `sqlalchemy.exc.DatabaseError`.  It's not super-precise, but it at least restricts to "you asked to do something and the DB said no".

**Pull request merged**
